# Understanding Your Data (Pandas Only) - Titanic Dataset

```text
1. Understanding Your Data
   │
   ├── Asking Basic Questions
   │
   ├── 1 How big is the data?
   │      → shape
   │      → number of rows
   │      → number of columns
   │
   ├── 2 How does the data look?
   │      → head()
   │      → tail()
   │      → sample()
   │
   ├── 3 What are the data types of columns?
   │      → dtypes
   │      → info()
   │
   ├── 4 Are there any missing values?
   │      → isnull()
   │      → isna()
   │      → missing-value counts - df.isnull().sum()
   |      → Remove missing value rows  - df=df.dropna() / df.dropna(inplace=True)
   │
   ├── 5 How does the data look mathematically?
   │      → describe() - numeric data only
   │      → mean
   │      → std
   │      → min
   │      → quartiles
   │      → max
   │
   ├── 6 Are there any duplicate values?
   │      → duplicated()
   │      → duplicated().sum()
   |      → drop duplicate rows - df=df.drop_duplicates() / df.drop_duplicates(inplace=True)
   │
   └── 7 What is the correlation between the columns?
          → corr() - numeric data only
          → correlation matrix
          → heatmap
```


In [26]:
import pandas as pd

In [45]:
df = pd.read_csv('train.csv')

## 1. How big is the data?

In [ ]:
# rows, cols
df.shape

(183, 12)

In [ ]:
# rows
len(df)

183


## 2. How does the data look?

In [ ]:
# returns data frame
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [ ]:
# random - returns data frame
df.sample(5)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
303,304,1,2,"Keane, Miss. Nora A",female,NaN,0,0,226593,12.3500,E101,Q
495,496,0,3,"Yousseff, Mr. Gerious",male,NaN,0,0,2627,14.4583,NaN,C
721,722,0,3,"Jensen, Mr. Svend Lauritz",male,17.0,1,0,350048,7.0542,NaN,S
116,117,0,3,"Connors, Mr. Patrick",male,70.5,0,0,370369,7.7500,NaN,Q
563,564,0,3,"Simmons, Mr. John",male,NaN,0,0,SOTON/OQ 392082,8.0500,NaN,S


In [31]:
df.tail()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.00,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.00,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.45,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.00,C148,C
890,891,0,3,"Dooley, Mr. Patrick",male,32.0,0,0,370376,7.75,NaN,Q


## 3. What are the data types of columns?

In [32]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name               str
Sex                str
Age            float64
SibSp            int64
Parch            int64
Ticket             str
Fare           float64
Cabin              str
Embarked           str
dtype: object

In [33]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 83.7 KB


## 4. Are there any missing values?

In [34]:
df.isna()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,False,False,False,False,False,False,False,False,False,False,True,False
1,False,False,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,True,False
3,False,False,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...
886,False,False,False,False,False,False,False,False,False,False,True,False
887,False,False,False,False,False,False,False,False,False,False,False,False
888,False,False,False,False,False,True,False,False,False,False,True,False
889,False,False,False,False,False,False,False,False,False,False,False,False


In [35]:
df.isnull()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,False,False,False,False,False,False,False,False,False,False,True,False
1,False,False,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,True,False
3,False,False,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...
886,False,False,False,False,False,False,False,False,False,False,True,False
887,False,False,False,False,False,False,False,False,False,False,False,False
888,False,False,False,False,False,True,False,False,False,False,True,False
889,False,False,False,False,False,False,False,False,False,False,False,False


In [36]:
# most used
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

# Dropping Columns Quick Notes

1. **Drop columns / rows**

   * `df.drop(...)` → Removes specified rows or columns.
   * `axis=1` → columns | `axis=0` → rows
   * `data.drop('col1', axis=1)` → Removes a **single column**.
   * `data.drop(['col1', 'col2'], axis=1)` → Removes **multiple columns**.
   * `data.drop('Unnamed: 0', axis=1)` → Removes an unwanted index column such as `0, 1, 2, 3...` accidentally saved in a CSV.

2. **Missing values**

   * `df.dropna()` → Removes rows containing **missing (`NaN`) values**.

3. **Duplicate rows**

   * `df.drop_duplicates()` → Removes **duplicate rows**.

4. **Returned DataFrame & `inplace`**

   * `df.drop(...)`, `df.dropna()`, `df.drop_duplicates()` → By default, **return a new DataFrame**; the original DataFrame is unchanged.
   * Assign back to save the change: `df = df.dropna()`
   * `inplace=True` → Modifies the **original DataFrame directly**.
   * Examples: `df.dropna(inplace=True)`, `df.drop_duplicates(inplace=True)`, `df.drop('col1', axis=1, inplace=True)`

   **Pandas `inplace=True`**

   Simple rule:

   If the Pandas method **supports** **`inplace=True`**, then:

   ```python
   df.method(inplace=True)
   ```

   modifies the original `df`.

   But **don't assume every Pandas method supports it**.

5. **Check remaining data**

   * `len(df)` → Returns the **number of rows** remaining in the DataFrame.


In [37]:
df.dropna()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
10,11,1,3,"Sandstrom, Miss. Marguerite Rut",female,4.0,1,1,PP 9549,16.7000,G6,S
11,12,1,1,"Bonnell, Miss. Elizabeth",female,58.0,0,0,113783,26.5500,C103,S
...,...,...,...,...,...,...,...,...,...,...,...,...
871,872,1,1,"Beckwith, Mrs. Richard Leonard (Sallie Monypeny)",female,47.0,1,1,11751,52.5542,D35,S
872,873,0,1,"Carlsson, Mr. Frans Olof",male,33.0,0,0,695,5.0000,B51 B53 B55,S
879,880,1,1,"Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)",female,56.0,0,1,11767,83.1583,C50,C
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S


In [38]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [ ]:
df = df.dropna()

In [44]:
df.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

In [49]:
df.dropna(inplace=True)
df.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

## 5. How does the data look mathematically?

In [39]:
# describes() - applies on numeric data only
df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


#### What to look for in describe()

- `count` → Compare with total rows → identify missing values.
- `Mean` → "Where is the center?"
- `mean vs 50% (median)` → Large difference may indicate skewness.
- `min/max` → Look for extreme or unusual values / possible outliers.
- `std` → Understand How spread out are the values around that center. That number came from looking at every value's distance from the mean
- `25% and 75%` → Understand where the middle 50% of data lies
- `min → 25% → 50% → 75% → max` → Get a quick idea of how the values are distributed

### Titanic insights

- `Age`: 714 vs 891 → 177 missing values
- `Fare`: mean (32.20) >> median (14.45) → likely right-skewed because some very high fares are pulling the mean upward.
- `Age`: mean (29.7) ≈ median (28) → no strong skew indicated.
- `Age`: min (0.42) to max (80) → recorded ages range from an infant to an 80-year-old passenger.
- `Fare`: 0 → 512.33 → wide range; investigate extreme values. Here fare seems fine
- min/max alone cannot prove something is an outlier. To actually determine outliers, we'd use something like the IQR method or a box plot later
- `Age`: mean = 29.7, std = 14.5 → average age is 29.7 years, with considerable variation in ages around this average.
- `Fare`: mean = 32.20, std = 49.69, while the central 50% lies only between 7.91–31 → fares are highly spread out, with some very high fares (max = 512.33) pulling the mean upward.
- `Age`: 25% = 20 and 75% = 38 → central 50% of recorded passenger ages lies between 20 and 38 years.
- `Age`: min → 25% → 50% → 75% → max = `0.42 → 20 → 28 → 38 → 80` → gives a quick picture of the age range, center, and where the central portion of the data is concentrated.

## 6. Are there any duplicate values?

In [50]:
df.duplicated()

1      False
3      False
6      False
10     False
11     False
       ...  
871    False
872    False
879    False
887    False
889    False
Length: 183, dtype: bool

In [51]:
df.duplicated().sum()

np.int64(0)

In [52]:
# df.drop_duplicates() returns df 
df = df.drop_duplicates()
len(df)

183

## 7. What is the correlation between the columns?

In [42]:
# corr() calculates the correlation between numerical columns. "Ignore non-numeric columns and calculate correlation only among the numeric columns
df.corr()

ValueError: could not convert string to float: 'Braund, Mr. Owen Harris'

In [ ]:
df.corr(numeric_only=True)

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
PassengerId,1.000000,-0.005007,-0.035144,0.036847,-0.057527,-0.001652,0.012658
Survived,-0.005007,1.000000,-0.338481,-0.077221,-0.035322,0.081629,0.257307
Pclass,-0.035144,-0.338481,1.000000,-0.369226,0.083081,0.018443,-0.549500
Age,0.036847,-0.077221,-0.369226,1.000000,-0.308247,-0.189119,0.096067
SibSp,-0.057527,-0.035322,0.083081,-0.308247,1.000000,0.414838,0.159651
Parch,-0.001652,0.081629,0.018443,-0.189119,0.414838,1.000000,0.216225
Fare,0.012658,0.257307,-0.549500,0.096067,0.159651,0.216225,1.000000


In [ ]:
# Output Column- y - Survived
df.corr(numeric_only=True)['Survived']

PassengerId   -0.005007
Survived       1.000000
Pclass        -0.338481
Age           -0.077221
SibSp         -0.035322
Parch          0.081629
Fare           0.257307
Name: Survived, dtype: float64

# Correlation

```text
-1                 0                 +1
│------------------│------------------│
strong negative    no linear          strong positive
                   relationship

Correlation:
- +1 → perfect positive linear relationship
-  0 → no linear relationship
- -1 → perfect negative linear relationship
- How columns are related with each other. How does one column affect other column linearly


#### Strong positive correlation → +1

Y
↑
|                    ●
|                ●
|            ●
|        ●
|    ●
| ●
+--------------------------→ X

- As X increases, Y also increases
- Hours studied ↑  →  Marks ↑
- upward straight-line pattern
- Correlation is positive and strong → close to +1.


#### Strong negative correlation → -1

Y
↑
| ●
|     ●
|         ●
|             ●
|                 ●
|                     ●
+--------------------------→ X

- As X increases, Y decreases
- Car age ↑  →  Car price ↓
- downward straight-line pattern
- Correlation is negative and strong → close to -1


#### No linear correlation → 0

Y
↑
|       ●          ●
|  ●             ●
|          ●
|     ●               ●
|               ●
| ●       ●
+--------------------------→ X

- As X increases, Y doesn't consistently increase or decrease.
- Hours studied → Shoe size
- no clear straight-line pattern
- Correlation is close to 0 → little/no linear relationship
- A feature with low/zero correlation with the target can be considered as a candidate for removal, but correlation alone should not be used as the final decision because correlation measures only linear relationships. The feature may still have a non-linear relationship with the target and therefore could still be useful for prediction.
```


### Insights Derived
- Fare ↑ and Survival ↑  → If one tends to increase when the other increases → positive correlation
- Pclass ↑ and Survival ↓ → If one tends to increase when the other decreases → negative correlation
- PassengerId and Survival → PassengerId is just an identifier, so its correlation with survival isn't meaningful for modeling